### ISSUE TITLE
Compile loop because output directory is being included

### ISSUE THREAD

**Issue by danielo515** [Microsoft/TypeScript#13823]
Hello,

I'm having problems using the watch option of typescript compiler. The output directory is being watched/compiled which in turn triggers another compilation, that triggers another compilation, you get the idea right? Each time a compilation is triggered a new sub-directory is created that contains all the compiled code.

**TypeScript Version:**  2.1.5

**Code**
Actually I think it is better if I specify the configuration that I'm using

```ts
{
    "compilerOptions": {
        "compileOnSave":true,
        "watch": true,
        "module": "commonjs",
        "target": "es5",
        "allowJs": true,
        "noImplicitAny": true,
        "sourceMap": true,
        "strictNullChecks": true,
        "outDir": "./dist",
        "rootDir": "./src"
    },
    "exclude": [
        "node_modules",
    ]
}
```

**Expected behavior:**

I want the compiler to generate compiled output to the dist directory and ignore its content.

**Actual behavior:**

As I explained, it is compiling the compiled code on an infinite loop.

I think that I can just put the `src` directory on the exclude array, but it is not documented on any place that the exclude array can be used for this purpose.

**Comment by mhegazy** [comment]
should not the exclude be:
```ts
    "exclude": [
        "./dist",
        "node_modules",
    ]
```

**Comment by danielo515** [comment]
Dear @mhegazy,

I also came to that conclusion, but it doesn't looks like a right solution because:

1. There is an option called outup dir. Shouldn't tsc ignore it by default ?
1. There is an option called rootDir. Shouldn't tsc process only the contents within that directory ?
1. The documentation about the exclude property is unclear , and seems that it only should be used for excluding node_modules and that kind of directories.

Checking the documentation makes me think stronger that this is a bug or an error on the documentation :

>Files in the directory specified using the "outDir" compiler option are always excluded unless explicitly included via the "files" property (even when the “exclude” property is specified).

As you can see ,my assumptions were correct according to the documentation,so something is wrong here.

In my opinion this is a bug or an unexpected behavior.

Here is the documentation link 
https://www.typescriptlang.org/docs/handbook/tsconfig-json.html

**Comment by mhegazy** [comment]
you have `allowJs` turned on. i.e. you are telling the compiler to pick up all .js files in your included files.  and your include/exclude pattern does not exclude the `outDir`. the compiler does not know what files it will generate, when it starts to look for files to process, since it needs the input to know what are the expected output. By default the exclude pattern includes `outDir`. but you overwrote that with one that only excludes `node_modules`. so options:
- remove the exclude pattern, or just add `outDir` your self
- add an include pattern to only include `./src/*`
- do not enable `allowJs` if you do not need it

**Comment by danielo515** [comment]
> you have allowJs turned on. i.e. you are telling the compiler to pick up all .js files in your included files.

I know that is what is causing it (can't call it "the problem"). But there is nothing on the documentation that specifies that you can not turn `allowJs` on and use the outDir,

> By default the exclude pattern includes outDir. but you overwrote that with one that only excludes node_modules

I expected those two to be merged, not exclusive. Is there any place on the documentation that specifies that the `outDir` is ignored as exclude if you use the `exclude` property ? If it does, I missed it.

> remove the exclude pattern, or just add outDir your self

If I just remove the exclude patter, will it pick up the node_modules folder? I guess the answer is yes.

> add an include pattern to only include ./src/*

That seems to be the best option, but the documentation is a bit confusing about how and when you should use the include pattern.

> do not enable allowJs if you do not need it

I have enabled JS because I want to use it. Maybe in a near future I could use just TS.

I'll try your suggestions and report back.
